packages/lodestar/src/chain/chain.ts

```diff
@@ -7,51 +7,55 @@ import BN from "bn.js";
 import {EventEmitter} from "events";
 import {hashTreeRoot} from "@chainsafe/ssz";
 
-import {
-  Attestation,
-  BeaconBlock,
-  BeaconState,
-  Deposit,
-  Eth1Data,
-  number64,
-  uint16,
-  uint64
-} from "../types";
+import {Attestation, BeaconBlock, BeaconState, number64, uint16, uint64} from "../types";
 import {DEPOSIT_CONTRACT_TREE_DEPTH, GENESIS_SLOT} from "../constants";
 
 import {IBeaconDb} from "../db";
 import {IEth1Notifier} from "../eth1";
 import {ILogger} from "../logger";
 import {IBeaconConfig} from "../config";
 
-import {getEmptyBlock, getGenesisBeaconState} from "./genesis";
+import {getEmptyBlock, initializeBeaconStateFromEth1, isValidGenesisState} from "./genesis/genesis";
 
 import {stateTransition} from "./stateTransition";
 
 import {LMDGHOST, StatefulDagLMDGHOST} from "./forkChoice";
 import {getAttestingIndices, slotToEpoch} from "./stateTransition/util";
 import {IBeaconChain} from "./interface";
-import {ProgressiveMerkleTree} from "../util/merkleTree/merkleTree";
+import {ProgressiveMerkleTree} from "../util/merkleTree";
 import {processSortedDeposits} from "../util/deposits";
+import {OpPool} from "../opPool";
+import {Block} from "ethers/providers";
+
+export interface IBeaconChainModules {
+  config: IBeaconConfig;
+  opPool: OpPool;
+  db: IBeaconDb;
+  eth1: IEth1Notifier;
+  logger: ILogger
+}
 
 export class BeaconChain extends EventEmitter implements IBeaconChain {
+
   public chain: string;
-  public genesisTime: number64;
+  public genesisTime: number64 = null;
   public forkChoice: LMDGHOST;
   public chainId: uint16;
   public networkId: uint64;
-  private config: IBeaconConfig;
+
+  private readonly config: IBeaconConfig;
   private db: IBeaconDb;
+  private opPool: OpPool;
   private eth1: IEth1Notifier;
-  private _latestBlock: BeaconBlock;
   private logger: ILogger;
 
-  public constructor(opts, {config, db, eth1, logger}: { config: IBeaconConfig; db: IBeaconDb; eth1: IEth1Notifier; logger: ILogger }) {
+  public constructor(opts, {config, db, eth1, opPool, logger}: IBeaconChainModules) {
     super();
     this.chain = opts.chain;
     this.config = config;
     this.db = db;
     this.eth1 = eth1;
+    this.opPool = opPool;
```

Comment:
I think this creates a circular dependency

Reply:
Oh, damn. Any idea how to proceed. All my ideas are bad in some way: 
- call opPool.removeOperations directly 
- take chain module as param in start method of opPool
- fetch deposits directly from database in chain module

I don't like any of it. Maybe to pass some OpPool in a Promise :/

Reply:
How is this circular?

Reply:
Eth1 was depending on OpPool, OpPool was depending on chain and chain was depending on opPool and eth1. No way to initialize them. Fixed it in latest commit